README.md
# kelly

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-06-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-06-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/kelly/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/kelly/compare/0.1.0...0.1.1


test/kelly/core_test.clj
(ns kelly.core-test
  (:require [clojure.test :refer :all]
            [kelly.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/kelly/ts.clj
(ns kelly.ts
  (:require
    [ts.csv :as csv]
    [clojure.pprint :refer [pprint]])
  (:import (java.time LocalDate)))


(defn parse-dollars
  [string]
  (int (* 100 (Double/parseDouble string))))

(defmacro defmemoized
  [fn-name args & body]
  `(def ~fn-name (memoize (fn ~fn-name ~args ~@body))))

(defmemoized parse-shiller-sp500 [{:keys [^Integer from-year ^Integer to-year]}]
  (->> (csv/parse
         "shiller-sp-500.csv"
         [{:id     :ts/date
           :column 0
           :parser #(LocalDate/parse %)}
          {:id     :ts/price
           :column 1
           :parser parse-dollars}
          {:id     :dividend
           :column 7
           :parser parse-dollars}])
       (filter (fn [ts]
                 (and (.isAfter (:ts/date ts) (LocalDate/of from-year 1 1))
                      (not (.isAfter (:ts/date ts) (LocalDate/of to-year 1 1))))))))

(defn period-to-period-price-deltas
  [ts {:keys [periods]}]
  (map (fn [left right]
         (double (/ (- (:ts/price right)
                       (:ts/price left))
                    (:ts/price left))))
       ts (drop periods ts)))

(defn average [nums]
  (double (/ (reduce + nums)
             (count nums))))





src/kelly/plot.clj
(ns kelly.plot
  (:require
    [medley.core :refer [map-vals]]
    [kelly.ts :as ts]))

(defn- get-schema
  [plot]
  (let [[_vega-light schema] plot]
    schema))

(defn- get-plot-values
  [plot]
  (-> (get-schema plot) :data :values))



(defn histogram
  [values & {:keys [width step title]
             :or {width 600
                  step 1
                  title "x"}}]

  [:vega-lite
   {:data {:values (map (fn [x] {:x x}) values)}
    :mark "bar"
    :title title
    :width width
    :encoding {:x {:field "x"
                   :title title
                   :type "quantitative"
                   :bin (if step {:step step}
                                 true)}
               :y {:type "quantitative"
                   :aggregate "count"}}}])

(defn join-histograms
  [histograms & {:keys [width step title]}]
  (let [field :title
        values
          (->> histograms
               (map (fn [plot]
                      (let [schema (get-schema plot)
                            vega-light-values (get-plot-values plot)
                            histogram-title (get schema field)]
                        (map #(assoc % field histogram-title) vega-light-values))))
             (apply concat))
        values-by-field
          (->> values
               (group-by field)
               (map-vals #(-> % :x ts/average)))]
    [:div
     [:div "Averages"
      [:ul
       (for [[field-name average] values-by-field]
         [:li field-name ": " average])]]
     [:vega-lite
      {:data {:values values}
       :mark "bar"
       :title title
       :width width
       :encoding
         {:row {:field (name field)
                :type "nominal"}
          :x {:field "x"
              :title title
              :type "quantitative"
              :bin (if step {:step step}
                            true)}
          :y {:type "quantitative"
              :aggregate "count"
              :stack false}
          :color {:type "nominal"
                  :field (name field)}
          :opacity {:value 0.5}}}]]))

(defn time-series
  [values & {:keys [width x y] :or {width 600}}]
  (assert (keyword? x))
  (assert (keyword? y))
  [:vega-lite
    {:width width
     :data {:values values}
     :mark {:type "line", :point true},
     :encoding {:x {:field (name x)
                    :type "quantitative"},
                :y {:field (name y)
                    :type "quantitative"}}}])


(defn values->time-series
  [values & {:keys [width x y]
             :or {width 600 x :step y :value}}]
  (-> (map-indexed
        (fn [index value]
         (hash-map x index y value))
        values)
      (time-series :x x :y y :width width)))


(defn format-% [number]
  (format "%.2f%%" (* 100 number)))


src/kelly/random.clj
(ns kelly.random
  (:require
    [medley.core :refer [map-vals]])
  (:import (java.util Random)))

(def random (new Random 0xc0de))



(defn toss-coin
  "Tosses a coin with the given probability of winning.

  Returns -1 for a loss and +1 for a win."
  [& {:keys [p when-win when-loss] :or {when-win true when-loss false}}]
  (let [limit 1e6
        cut-off (* p limit)
        rnd (.nextInt random limit)]

    (if (<= rnd cut-off)
      when-win
      when-loss)))

(defn create-single-value-history-store
  []
  {:initial identity
   :current-value identity
   :appender (fn [_ new-value] new-value)})

(defn create-vector-history-store
  []
  {:initial vector
   :current-value peek
   :appender conj})

(let [n 1e1]
  (->> (map (fn [_] (toss-coin :p 0.5)) (range))
       (take n)
       (group-by identity)
       ;(map-vals #(/ (count %) n))
       (println)))

src/kelly/sims/coin_toss.clj
(ns kelly.sims.coin-toss
  (:require
    [kelly.random :as rnd]
    [kelly.preview :refer [preview]]
    [medley.core :refer [map-vals]]
    [ts.plot :as plot]))

(defn simulate-biased-coin
  [{:keys [edge minimum-bet starting-cash bet-size stop-at]}]
  (loop [history [starting-cash]]

    (let [remaining-cash (peek history)
          bet-amount (* bet-size remaining-cash)]

      (if (or (<= remaining-cash minimum-bet)
              (>= remaining-cash stop-at)
              (> (count history) 1000))

        history

        (let [amount-if-won (+ remaining-cash bet-amount)
              amount-if-loss (- remaining-cash bet-amount)]
          (recur
            (conj history
                  (rnd/toss-coin
                    :p (+ 0.5 edge)
                    :when-win amount-if-won
                    :when-loss amount-if-loss))))))))





(defn format-percentage [number]
  (format "%.2f%%" (* 100 number)))


(defn view-biased-coin-simulation
  [{:keys [edge starting-cash bet-size stop-at minimum-bet]}]
  (let [run-sim
        #(simulate-biased-coin
            {:edge edge
             :minimum-bet minimum-bet
             :starting-cash starting-cash
             :bet-size bet-size
             :stop-at stop-at})]
    [:div
     [:ul
      [:li "Edge: " (format-percentage edge)]
      [:li "Starting Cash: " starting-cash]
      [:li "Bet Size: " (format-percentage bet-size) " of remaining cash"]
      [:li "Stop At: " stop-at]]

     [:vega-lite (plot/values->time-series (run-sim) :x :step :y :cash)]

     [:vega-lite (plot/histogram (->> (repeatedly 1000 run-sim)
                                      (map #(-> % peek)))
                                 :title "Remaining Cash"
                                 :width 600
                                 :step (/ starting-cash 2))]]))




(defn render-main []
  (let [edge 0.05
        minimum-bet 100
        starting-cash 1e4
        stop-at 1e5]
    [:div
      [:h1 "Explorations on the Kelly Criterion"]

      [:p "The Kelly Criterion is an attempt to solve the question of what fraction of your total capital should be put at risk in order to maximize long term survival and capital growth."]

      [:h2 "Simplest Kelly"]
      [:p "Let us begin our exploration of the Kelly Criterion with the most basic formulation."]

      [:p "Imagine there is an (oddly generous) casino that offers a bet with a " (format-percentage (+ 0.5 edge)) " odds of doubling, meaning you have an edge of " (format-percentage edge) ". A positive edge means that you can make money in the long term thanks to the law of large numbers, but how can you use this edge to maximise your profits while avoiding ruin?"]

     [:p "Let's try to explore this problem empirically, by running monte carlo simulations. We will assume that the player starts with a fixed amount and will stop once he gets to the goal or goes bankrupt."

      [:h3 "Attempt #1: bet 50%"]
      [:p "Betting all your money seems risky already from the start, but lets run a simulation to see how well we can do."]
      (view-biased-coin-simulation {:edge edge
                                    :minimum-bet minimum-bet
                                    :starting-cash starting-cash
                                    :bet-size 0.5
                                    :stop-at stop-at})

      [:p "Not surprisingly it seems to always end up in ruin."]

      [:h3 "Attempt #2: bet 10%"]
      [:p "Let's try a more conservative approach now, but betting 10% instead."]
      (view-biased-coin-simulation {:edge edge
                                    :minimum-bet minimum-bet
                                    :starting-cash starting-cash
                                    :bet-size 0.10
                                    :stop-at stop-at})
      [:p "If you run this simulation a few times you will notice that its quite volatile. Sometimes we do end up at our goal but very often it ends up in ruin."]

      [:h3 "Attempt #3: bet 5%"]
      [:p "This option seems to consistently make decent profits."]
      (view-biased-coin-simulation {:edge edge
                                    :minimum-bet minimum-bet
                                    :starting-cash starting-cash
                                    :bet-size 0.05
                                    :stop-at stop-at})

      [:h3 "Attempt #4: bet 2.5%"]
      [:p "This option seems to consistently make decent profits."]
      (view-biased-coin-simulation {:edge edge
                                    :minimum-bet minimum-bet
                                    :starting-cash starting-cash
                                    :bet-size 0.025
                                    :stop-at stop-at})

      [:h3 "Attempt #5: bet 1%"]
      [:p "It doesn't end up in ruin, but it also never reaches our goal. Safe but unprofitable."]
      (view-biased-coin-simulation {:edge edge
                                    :minimum-bet minimum-bet
                                    :starting-cash starting-cash
                                    :bet-size 0.01
                                    :stop-at stop-at})

      [:h2 "What does Kelly Recommend?"]

      [:p "Kelly's formula produces the optimal wager at every bet and can be summarized by edge/odds. Applying this formula to our example we get edge=0.05"]]]))


(preview (render-main))

src/kelly/sims/sp500.clj
(ns kelly.sims.sp500
  (:require
    [kelly.random :as rnd]
    [ts.plot :as plot]
    [ts.finance :refer [future-value]]
    [ts.random :refer [random-float]]
    [kelly.ts :as ts]
    [kelly.preview :refer [preview]]
    [medley.core :refer [map-vals]]))

(defn repeatedly-fast
  [^Integer n func]
  (loop [^Integer n n
         result (transient [])]
    (if (<= n 0)
      (persistent! result)
      (recur
        (dec n)
        (conj! result (func))))))


(defn add-black-swans
  "Reads a number of tail events from `numbers` and based on that tail sample,
  it will return a new distribution with additional tail events as specified by
  `num-tail-events-to-add`."
  [num-tail-events-sample num-tail-events-to-add numbers]

  (let [sorted-numbers (sort numbers)
        tail (take num-tail-events-sample sorted-numbers)
        tail-generator! #(random-float (first tail) (last tail))]

    (->> (repeatedly-fast num-tail-events-to-add tail-generator!)
         (into numbers))))


(defn plot-sp500-histogram
  [{:keys [period-to-period-length
           from-year
           to-year]}]
  (let [ts (ts/parse-shiller-sp500 {:from-year from-year :to-year to-year})
        percentages (->> (ts/period-to-period-price-deltas ts {:periods period-to-period-length})
                         (add-black-swans 5 5)
                         (map #(* 100 %)))
        expected-value (ts/average percentages)]
    [:div
     [:vega-lite (plot/histogram percentages
                   :width 600
                   :step 1
                   :title (str "SP500 " period-to-period-length "-month returns from " from-year " - " to-year))]
     [:p "Expected Value: " expected-value "%"]]))


(defn run-sp500-kelly-simulation
  [{:keys [^Double fraction history-store initial-cash]}]

  (let [
        create-initial-history (:initial history-store)
        history-store-appender (:appender history-store)
        history-store-current-value (:current-value history-store)
        period-to-period-length 1
        years-to-invest 20
        transaction-cost 5
        total-periods-to-invest (* (/ 12 period-to-period-length) years-to-invest)
        ts (ts/parse-shiller-sp500 {:from-year 1990 :to-year 2007})
        deltas (->> (ts/period-to-period-price-deltas ts {:periods period-to-period-length})
                    (add-black-swans 2 5))]


    (loop [history (create-initial-history initial-cash)
           step 0]
      (if (or (>= step total-periods-to-invest)
              (<= (history-store-current-value history)
                  (* 0.7 initial-cash)))
        history

        (let [current-cash (double (history-store-current-value history))
              delta (double (rand-nth deltas))
              cash-to-invest-pre (* fraction current-cash)
              cash-to-invest-post (* (+ 1 delta) cash-to-invest-pre)
              static-cash (- current-cash cash-to-invest-pre transaction-cost)
              resulting-cash (+ static-cash (* cash-to-invest-post))]

          (recur (history-store-appender history (double resulting-cash))
                 (inc step)))))))

(defn probability-less-than-value
  [value points]
  (float (/ (->> points
                 (filter #(<= % value))
                 (count))
            (count points))))

(defn probability-greater-than-value
  [value points]
  (float (/ (->> points
                 (filter #(>= % value))
                 (count))
            (count points))))




(defn run-find-optimal-fraction []
  (->> (range 0.05 1 0.10)
       (pmap (fn [fraction]
               (println "Processing:" fraction)
               (let [initial-cash 10000
                     run-sim #(run-sp500-kelly-simulation
                                {:fraction fraction
                                 :history-store (rnd/create-single-value-history-store)
                                 :initial-cash initial-cash})
                     num-iterations 1000
                     resulting-cashes (repeatedly-fast num-iterations run-sim)]
                 (println "Done with: " fraction)
                 {:probability-of-market-returns
                    (probability-greater-than-value
                      (future-value {:present-value initial-cash
                                     :interest-rate 0.04
                                     :periods 20})
                      resulting-cashes)
                  :probability-of-worse-than-initial-cash
                    (probability-less-than-value
                      initial-cash
                      resulting-cashes)
                  :probability-of-worse-than-t-bill
                    (probability-less-than-value
                      (future-value {:present-value initial-cash
                                     :interest-rate 0.025
                                     :periods 20})
                      resulting-cashes)
                  :fraction fraction})))
       (into [])))

(defn plot-sp500-kelly-simulation
  [{:keys [fraction]}]
  (let [initial-cash 10000
        run-sim #(run-sp500-kelly-simulation
                   {:fraction fraction
                    :history-store %
                    :initial-cash initial-cash})
        points-in-histogram 1000
        resulting-cashes (->> #(run-sim (rnd/create-single-value-history-store))
                             (repeatedly-fast points-in-histogram)
                             (vec))
        t-bill-returns (future-value {:present-value initial-cash
                                      :interest-rate 0.025
                                      :periods 20})]
    [:div
     [:h2 "Fraction: " (plot/format-% fraction)]
     [:vega-lite (plot/values->time-series
                   (run-sim (rnd/create-vector-history-store))
                   :x :step
                   :y :cash)]

     [:p "Distribution of Returns"]
     [:vega-lite (plot/histogram
                   resulting-cashes
                   :step 2000
                   :title "Returns")]

     [:p "Probability of doing worse than the initial investment amount"]]
    (let [optimal-fraction-results (run-find-optimal-fraction)]
      [:div
       [:vega-lite
        (plot/time-series
          optimal-fraction-results
          :x :fraction
          :y :probability-of-market-returns)]
       [:vega-lite
        (plot/time-series
          optimal-fraction-results
          :x :fraction
          :y :probability-of-worse-than-initial-cash)]
       [:vega-lite
        (plot/time-series
          optimal-fraction-results
          :x :fraction
          :y :probability-of-worse-than-t-bill)]])))







(defn render-main []
  (let [edge 0.03]
    [:div
     [:h1 "Explorations on the Kelly Criterion: SP500"]

     [:p "Let's continue now our exploration of the Kelly Criterion with a more practical application: the returns of the SP500"]
     [:p "In order to create a proper simulation we will need to simulate some sort of bet "]

     (let [period-to-period-length 1
           initial-year 1965]
       [:div
         [:h2 "Estimating an Edge"]
         [:p "Let's first examine the distribution of the SP500's " period-to-period-length "-month returns, by looking at three distinct periods."]
         (plot-sp500-histogram {:from-year 1990
                                :to-year 2007
                                :period-to-period-length period-to-period-length})
         #_(plot-sp500-histogram {:from-year (+ initial-year 10)
                                   :to-year (+ initial-year 30)
                                   :period-to-period-length period-to-period-length})
         #_(plot-sp500-histogram {:from-year (+ initial-year 20)
                                    :to-year (+ initial-year 40)
                                    :period-to-period-length period-to-period-length})

         [:p "From this limited dataset we can see that there does seem to be an edge in the markets, which is related to the whole idea of index investing being the 'best' form of investment. We can also see that the edge seems to be consistently positive."]

         [:h2 "The Experiment"]
         [:p "Let's setup our experiment as follows:"]
         [:ul
          [:li "Every round we take a random " period-to-period-length "-month return from the distribution. "]
          [:li "If the return is positive we 'sell' our position and buy again, thus increasing our exposure by re-balancing"]
          [:li "If the return is negative we 'sell' our position and buy again, thus decreasing our exposure by re-balancing"]

          (plot-sp500-kelly-simulation {:fraction 0.6})]])]))



(let [result (time (render-main))]
  (println "Sending off to preview...")
  (preview result))